### PR TITLE
🏛️ 🔑 ⛓️ 

### PR DESCRIPTION
**Pull request by estrattonbailey**
# Testing
There's a TestFlight you can install. App should behave exactly the same. Check that private values still require auth, try creating a new wallet, signing a WC transaction, etc.

I have an APK as well, will post in Slack.

# README.md from the `@/keychain` directory

A wrapper around `react-native-keychain` that provides better type safety and
error handling. It smooths over the core library's polymorphic return type and
thrown exceptions by implementing a discriminated union and pre-defined error
codes that are backwards compatible with our pre-existing keychain wrapper.

**Instead of using `react-native-keychain` directly, use this module.** That way we
can continue to improve our interface independent of the base library, and if we
ever need to swap out the base library in the future, we can more easily do
that. It will also help in testing because we can mock this module more easily
than the base library.

## Usage

The API is essentially the same as our pre-existing keychain wrapper, except
the return type.

```typescript
import * as keychain from '@/keychain';

await keychain.set(
  'my-keychain-key',
  'value',
  await keychain.getPrivateAccessControlOptions()
);
const bool = await keychain.has('my-keychain-key');
const result = await keychain.get('my-keychain-key');

await keychain.setObject('my-object', { foo: 'bar' });
const result = await keychain.getObject('my-object');

await keychain.setSharedWebCredentials('username', 'password');
const result = await keychain.getSharedWebCredentials('username', 'password');

const biometryType = await keychain.getSupportedBiometryType();

await keychain.remove('my-keychain-key');
await keychain.remove('my-object');

await keychain.clear();
```

The returned result type is a discriminated union, which allows for type-safe
access of values or errors i.e. no more `?.` coalescing for values. Error
handling is also codified, and this wrapper _will not throw._

```typescript
const { value, error } = await keychain.getObject('my-object');

if (value) {
   console.log(value.foo) // safe property access
} else {
   switch(error) {
      case(keychain.ErrorType.Unavailable):
      case(keychain.ErrorType.UserCanceled):
      case(keychain.ErrorType.NotAuthenticated):
         console.log(...)
         break;
      case(keychain.ErrorType.Unknown):
      default:
         console.log(`An unknown error occurred: ${error}`)
   }
}
```

### CODE REVIEW

**Review by jkadamczyk — COMMENTED** [review]
I love it
Code quality wise it's great, I will run a bunch of manual test myself, but we need very heavy QA for that for sure

File: src/keychain/index.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+}
+
+export async function set(
+  key: string,
+  value: string,
+  options: Options = {}
+): Promise<void> {
+  logger.debug(`keychain: set`, { key }, logger.DebugContext.keychain);
+
+  // only save public data to mmkv
```
Comment (brunobar79) [inline src/keychain/index.ts:121]: maybe it's worth mentioning somewhere in the readme that the keychain wrapper is currently a Frankenstein between native keychain and mmkv  and why.

CC: @skylarbarrera

File: src/model/wallet.ts [hunk 2]
```diff
@@ -32,7 +31,8 @@ import profileUtils, {
   addressHashedColorIndex,
   addressHashedEmoji,
 } from '../utils/profileUtils';
-import * as keychain from './keychain';
+import * as keychain from '@/model/keychain';
+import * as kc from '@/keychain';
```
Comment (brunobar79) [inline src/model/wallet.ts:35]: I think it would be less confusing  to just import `getSupportedBiometryType`

File: src/keychain/index.ts (hunk 1 above)
Comment (estrattonbailey) [reply src/keychain/index.ts:121]: Good call 👍

**Review by skylarbarrera — APPROVED** [review]
prob should have a QA pass but fire fire 🔥

**Comment by estrattonbailey** [comment]
Talked to Bruno and we should be good to go here 👍

**Pull request merged**